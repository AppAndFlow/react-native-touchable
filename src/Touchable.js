// @flow

import * as React from 'react';
import {
  Platform,
  TouchableOpacity,
  TouchableHighlight,
  TouchableWithoutFeedback,
  TouchableNativeFeedback,
  View,
  StyleSheet,
} from 'react-native';

// Touchable highlight is tricky and certain style props must be included
// either on the Touchable component or the inner container view. Some must also
// be included on both.
const OUTER_VIEW_KEYS = new Set([
  'margin',
  'marginHorizontal',
  'marginVertical',
  'marginLeft',
  'marginRight',
  'marginTop',
  'marginBottom',
]);
const BOTH_VIEW_KEYS = new Set([
  'borderRadius',
  'borderTopLeftRadius',
  'borderTopRightRadius',
  'borderBottomLeftRadius',
  'borderBottomRightRadius',
]);

function separateStyle(style) {
  if (!style) {
    return { inner: style, outer: null };
  }

  const resolvedStyle = StyleSheet.flatten(style);
  if (!resolvedStyle) {
    return { inner: null, outer: null };
  }
  const inner = {};
  const outer = {};

  Object.keys(resolvedStyle).forEach(key => {
    if (BOTH_VIEW_KEYS.has(key)) {
      outer[key] = resolvedStyle[key];
      inner[key] = resolvedStyle[key];
    } else if (OUTER_VIEW_KEYS.has(key)) {
      outer[key] = resolvedStyle[key];
    } else {
      inner[key] = resolvedStyle[key];
    }
  });

  return { inner, outer };
}

type Props = {
  feedback: 'opacity' | 'highlight' | 'none',
  native?: boolean,
  overflow?: boolean,
  onPress?: () => any,
  onLongPress?: () => any,
  disabled?: boolean,
  hitSlop?: { top?: number, bottom?: number, left?: number, right?: number },
  tintColor?: string,
  style?: any,
  children?: React.Node,
};

export default class Touchable extends React.Component<Props> {
  static defaultProps = {
    native: true,
    overflow: false,
    disabled: false,
  };

  render() {
    const {
      feedback,
      native,
      overflow,
      onPress,
      onLongPress,
      disabled,
      hitSlop,
      tintColor,
      ...others
    } = this.props;
    if (native && Platform.OS === 'android' && Platform.Version >= 21) {
      if (disabled) {
        return <View {...others} />;
      }
      return (
        <TouchableNativeFeedback
          // eslint-disable-next-line babel/new-cap
          background={TouchableNativeFeedback.Ripple(tintColor, overflow)}
          disabled={disabled}
          hitSlop={hitSlop}
          onPress={onPress}
          onLongPress={onLongPress}
        >
          <View {...others} />
        </TouchableNativeFeedback>
      );
    } else if (feedback === 'opacity') {
      return (
        <TouchableOpacity
          {...others}
          onPress={onPress}
          onLongPress={onLongPress}
          disabled={disabled}
          hitSlop={hitSlop}
        />
      );
    } else if (feedback === 'highlight') {
      const { style, children, ...othersWithoutStyle } = others;
      const { inner, outer } = separateStyle(style);
      return (
        <TouchableHighlight
          onPress={onPress}
          onLongPress={onLongPress}
          disabled={disabled}
          hitSlop={hitSlop}
          {...othersWithoutStyle}
          style={outer}
        >
          <View style={inner}>
            {children}
          </View>
        </TouchableHighlight>
      );
    } else if (feedback === 'none') {
      return (
        <TouchableWithoutFeedback
          disabled={disabled}
          hitSlop={hitSlop}
          onPress={onPress}
          onLongPress={onLongPress}
        >
          <View {...others} />
        </TouchableWithoutFeedback>
      );
    }
    throw new Error('Invalid feedback type');
  }
}
