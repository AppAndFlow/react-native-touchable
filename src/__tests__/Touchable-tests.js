// @flow

import * as React from 'react';
import { Platform, View } from 'react-native';
import renderer from 'react-test-renderer';

import Touchable from '../Touchable';

// TODO: remove this when Blob is properly mocked.
jest.mock('Blob', () => ({})).mock('URL', () => ({}));

const FEEDBACKS = ['opacity', 'highlight', 'none'];

function createTestForAllFeedbacks(spec, test) {
  FEEDBACKS.forEach(feedback => {
    it(`${spec} with ${feedback} feedback`, () => {
      test(feedback);
    });
  });

  it(`${spec} with native feedback`, () => {
    Platform.OS = 'android';
    test('none');
    Platform.OS = 'ios';
  });
}

describe('Touchable', () => {
  createTestForAllFeedbacks('renders correctly', feedback => {
    const component = renderer
      .create(
        <Touchable feedback={feedback}>
          <View />
        </Touchable>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  createTestForAllFeedbacks('supports multiple children', feedback => {
    const component = renderer
      .create(
        <Touchable feedback={feedback}>
          <View />
          <View />
          <View />
          <View />
        </Touchable>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  createTestForAllFeedbacks('passes style properly', feedback => {
    const component = renderer
      .create(
        <Touchable feedback={feedback} style={{ flex: 1 }}>
          <View />
        </Touchable>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
