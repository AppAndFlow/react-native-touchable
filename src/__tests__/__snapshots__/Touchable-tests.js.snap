// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Touchable passes style properly with highlight feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      Object {},
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View />
  </View>
</View>
`;

exports[`Touchable passes style properly with native feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
  testID={undefined}
>
  <View />
</View>
`;

exports[`Touchable passes style properly with none feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
  testID={undefined}
>
  <View />
</View>
`;

exports[`Touchable passes style properly with opacity feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View />
</View>
`;

exports[`Touchable renders correctly with highlight feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      null,
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={undefined}
  >
    <View />
  </View>
</View>
`;

exports[`Touchable renders correctly with native feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={undefined}
  testID={undefined}
>
  <View />
</View>
`;

exports[`Touchable renders correctly with none feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={undefined}
  testID={undefined}
>
  <View />
</View>
`;

exports[`Touchable renders correctly with opacity feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View />
</View>
`;

exports[`Touchable supports multiple children with highlight feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      null,
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={undefined}
  >
    <View />
    <View />
    <View />
    <View />
  </View>
</View>
`;

exports[`Touchable supports multiple children with native feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={undefined}
  testID={undefined}
>
  <View />
  <View />
  <View />
  <View />
</View>
`;

exports[`Touchable supports multiple children with none feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={undefined}
  testID={undefined}
>
  <View />
  <View />
  <View />
  <View />
</View>
`;

exports[`Touchable supports multiple children with opacity feedback 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View />
  <View />
  <View />
  <View />
</View>
`;
